plugins/inputs/modbus/modbus.go

```diff
@@ -159,6 +170,14 @@ func (m *Modbus) Init() error {
 		return fmt.Errorf("device name is empty")
 	}
 
+	if m.Retries < 0 {
+		return fmt.Errorf("retries cannot be negative")
+	}
+
+	if m.Retries > 0 && m.RetriesWaitTime.Duration < 0 {
```

Comment:
Duration needs to be positive.  0 wait time makes no sense

Reply:
I would agree this check is not adding any value. While it may technically be possible to set a negative duration, Go will not blow up and will interpret it as a zero wait. This check is too defensive and should probably be removed.

Reply:
@reimda:
> Duration needs to be positive. 0 wait time makes no sense
Well zero-wait time actually makes sense as it means to directly try again. Some devices take quite some time to respond and you can basically try to poll them in a busy-waiting fashion.

@ssoroka:
Well I just thought that a negative value is with 99.99999% probability a configuration error and instead of silently using another value (as Go will map it basically to 0) might hide some issues. Imagine you wanted to give e.g. 100ms but due to some cat running over your keyboard (ok this is made up) you get -100ms. You will wonder forever why this device will fail even though you thought you allowed 5 retries with 100ms wait time.

But anyway, I will follow the final decision as I interpret the above as @reimda wants a more strict check while @ssoroka  wants the check removed. So which way I should follow?

Reply:
Removed check as requested in commit below.

Reply:
@srebhan 
I initially thought the zero wait time was a simple off by one error.  Thanks for explaining how quick retries can be useful for devices that take a long time to reply that they're busy.

I'm new to modbus so maybe you can help me understand what's going on here.  If a device replies quickly that it's busy, won't this code immediately retry?  Without any delay it will be unlikely that the retries get anything other than more quick busy responses.

It seems to me that we need to handle both cases - retry quickly on slow busy responses and retry not too quickly on quick busy responses.

Maybe change this:
`time.Sleep(m.RetriesWaitTime.Duration)`
to this:
`elapsed = now() - timestamp
time.Sleep(m.RetriesWaitTime.Duration - elapsed)`

Then also have a small lower limit on duration, maybe 20ms

What do you think?

Reply:
I think you are over engineering here. Normally you know how long your device will likely be busy (which is usually only the case if some other process also querying the device) and how quickly it responds. Take for example my use-case, I have another process (that I cannot influence) querying the device and it takes that process about 500-700ms to finish. So I will probably set the retries to 5 and the wait time to 200ms to achieve a trade-off between the delay and the "guarantee" that one of the trials will succeed. Anyway in some other use-case the delay/latency of getting the information might be crucial, so in this case you probably do not set a wait interval but set the number of retries high.
This being said, it more depends on the use-case, the individual setup and last but not least on the individual preferences to choose between long-wait, less retries vs. busy-waiting with many retries... So I think the extra effort is not worth it. :-)